README.md
Online 15Puzzle simulator 在线15Puzzle模拟器
==========
- A simple (and crude) online 15p with settings and sessions.
- 一个简陋的在线15p, 带有设置和成绩分组功能.

Introduction 简介
----------
- This online 15Puzzle simulator is a sliding puzzle simulator with professional functions, which provides:
- variable puzzle size: from 2x2 to 20x20, unequal width and height is available
- plenty of customizable settings: modes, appearance, statistics, timer, etc.
- local storage for results and settings
- This simulator can be viewed on most of the modern browsers, on mobile devices or PC.

- 这是一个在线15Puzzle模拟器, 带有如下专业功能:
- 可变的大小: 从2x2到20x20, 支持不等阶
- 众多的设置: 模式, 外观, 统计, 计时器等
- 成绩和设置的本地存储
- 此模拟器可以用电脑或移动设备的多数浏览器浏览.

General 通用
----------
- How to play: Press the spacebar once, or click the Scramble button to scramble. You can also enter your customized scramble. Use a keyboard, a mouse, or touchscreen to drag around the pieces, until all of them are at their beggining states.

- Puzzle (main) page - The page for main gameplay.
- Result page - The page for viewing results and doing statistics.
- Setting page - The page for changing a variety of settings.

- 如何玩: 按一次空格, 或点击打乱按钮来打乱. 也可以输入自定义打乱. 用键盘, 鼠标, 或触屏来拖拽方块, 直到它们都回到初始位置.

- 方块 (主) 页面 - 在这里执行复原.
- 成绩页面 - 记录成绩, 统计成绩.
- 设置页面 - 改变设置.

Controls 控制
----------
Keyboard:
- (Left, Up, Right, Down) Arrow keys or (D, K, F, J) letter keys - control the movement of pieces. Invert control is supported.
- Spacebar - scramble, or confirm solve (blindfolded), or switch to next scramble (multi blindfolded).
- Esc key - cancel current attempt, or revert to solved state.
- "+", "-" keys - increase (or decrease) puzzle width and height by 1.
- PgUp, PgDn keys - increase (or decrease) puzzle width by 1.
- Home, End keys - increase (or decrease) puzzle height by 1.

Mouse or touchscreen:
- Hover is on - drag the mouse or a finger over a piece, and the piece(s) move over.
- Hover is off - click on a piece, and corresponding piece(s) move over.

键盘:
- (左上右下) 方向键或 (D, K, F, J) - 控制方块移动. 支持反向控制.
- 空格 - 打乱, 或确认复原 (盲拧), 或切换到下一个 (多盲).
- Esc - 自动复原, 取消本次还原.
- "+", "-" - 宽度和高度均+1 (-1).
- PgUp, PgDn - 宽度+1 (-1).
- Home, End - 高度+1 (-1).

鼠标或触屏:
- 滑动开 - 鼠标或手指划过方块来操作.
- 滑动关 - 点击方块操作.

Solve types 复原模式
----------
- Standard - do one solve with infinite inspection time.
- Relays - relay from biggest puzzle to the smallest puzzle.
- Marathon - do many solves in the same puzzle size.
- Blindfolded - do one solve with inspection peroid counted. Once inspection is finished (i.e. the first move is done), the puzzle is invisible until you confirm your solve.
- Multi Blindfolded - do many solves with inspection peroid counted. Memorize these puzzles at the same time, and solve them continuously. During inspection or solving, you can go to the previous puzzle or the next puzzle.

- 标准 - 复原一次, 不限观察时间.
- (阶数变化) 连拧 - 从最大的方块到最小, 连续复原.
- (阶数不变) 连拧 - 对同样大小的方块作多次复原.
- 盲拧 - 复原一次, 计观察时间. 观察一旦结束 (即动了第一步), 方块不可见, 直到确认复原.
- 多盲 - 复原多次, 计观察时间. 同时记下所有方块状态, 随后连续复原. 在观察或复原阶段, 可以向前或向后切换.

Result page 成绩页面
----------
This page includes functions below:
- switch between 15 sessions
- change the session name
- clear the session
- click on the # order to delete single solve
- click on single/average results to get detailed info
- click on session mean to get overall stats

这个页面有如下功能:
- 在15个分组之间切换
- 重命名分组
- 清空分组
- 点击#序号删除单次成绩
- 点击单次/平均成绩获取详细信息
- 点击分组平均获取整体统计

Color schemes 配色方案
----------
- Fringes - separate a puzzle with groups of one column conbined with one row. Recommeded for most players.
- Rows, columns - separate rows and columns alternatively. Recommeded for most players.
- Rows/Columns - separate into many rows/columns.
- Mono - the puzzle is in single color.
- Last 2 Rows/Columns - separate into rows/columns until the last 2 rows/columns. The last 2 - rows/columns are separated into many columns or rows. Recommended for row-by-row or column-by-column players.
- Live - highlight the current group and the next group. Especially recommended for big puzzles.

- 降阶 - 将整个方块分为很多阶 (一阶为一行一列). 对大多数人推荐.
- 行列交替 - 分为一行一列交替的配色. 对大多数人推荐.
- 行/列 - 分为很多行/列. 对层先/列先玩家推荐.
- 单色 - 整个方块的颜色相同.
- 按行, 最后2行按列/按列, 最后2列按行 - 分为行/列, 直到最后2行/列. 最后2行/列按列/行区分.
- 动态 - 高亮当前和下一组. 做高阶时尤其推荐.

CHANGELOG.md
Changelog 变更记录
==========
All changes to this project will be listed here.
对这个项目的所有变更都会在这里记录.

4.1.7 - 2022-02-21
----------
Added 添加
- Summary on main page
- 在主页面的小结

4.1.6 - 2022-02-19
----------
Added 添加
- function button (works as spacebar) for touchscreen
- 为触屏 (功能同空格) 添加的功能键

Changed 变更
- puzzle center aligned for inequal puzzles
- fixed keyboard control while multibld inspection
- fixed timer doesn't start in relay when scrambled for multiple times
- fixed scramble output
- 不等阶横向居中显示
- 修复多盲观察时键盘操作
- 修复连拧下多次打乱计时器不启动
- 修复打乱输出

4.1.5 - 2022-02-16
----------
Added 添加
- add sessions
- delete sessions
- 添加分组
- 删除分组

4.1.4 - 2022-02-15
----------
Changed 变更
- fixed a few minor bugs
- improved checking efficiency of live schemes 
- 修正一些小错误
- 提升动态配色的检测效率

4.1.3 - 2022-02-14
----------
Added 添加
- DNF in stats
- session overall stats
- 统计中设DNF
- 整个分组的统计信息

Changed 变更
- fixed bugs of result page
- 修正一些成绩统计的问题

4.1.2 - 2022-02-13
----------
Added 添加
- a few functions of result page
- 成绩界面的一些功能

Changed 变更
- fixed a few bugs on result statistics and puzzle base
- 修正一些成绩统计和进制的问题

4.1.1 - 2022-02-13
----------
Added 添加
- description files
- 说明文件

Changed 变更
- functionality of touchmove and buttons
- 触屏滑动和按钮的功能性

4.0 - before 2022-02-12
----------
Added 添加
- result page, and related statistics settings
- language
- timer settings
- 成绩页面, 以及相关的统计设置
- 语言
- 计时器设置

Changed 变更
- some changes on settings interface
- 设置页面一些变化

Removed 移除
- some rainbow settings
- 一些彩虹色设置

language.js
let language = 0; // 0 english, 1 chinese 2?
let buttonLanguageList = [
    ["Customize", "Scramble", "Solve", "Confirm solve", "Next puzzle", "Settings", "Puzzle", "Export state", "Previous puzzle", "Result", "Clear session", "Rename", "New", "Delete"],
    ["自定义", "打乱", "自动复原", "确认复原", "下一个", "设置", "方块", "导出状态", "上一个", "成绩", "清空分组", "重命名", "新建", "删除"]
];
let settingsTitleList = [
    ["Puzzle", "Appearance", "Statistics", "Color scheme", "Controls", "Timer", "General", "Session"],
    ["方块", "外观", "统计", "配色", "控制", "计时器", "通用", "分组"]
];
let settingsLanguageList = [
    ["Puzzle width (2 ~ 20)", "Puzzle height (2 ~ 20)", "Solve type", "Marathon solves (> 1)", "Multiblind solves (> 1)", "Text color",
    "Text size ratio (0 ~ 1)", "Puzzle size (> 0)", "Margin (>= 0)", "Fillet radius ratio (0 ~ 0.5)", "Border of pieces", "Base (2 ~ 36)",
    "Color scheme", "Rainbow lightness",
    "Live pieces: done", "current", "next", "other", "Invert keyboard controls", "Hover on", "Language",
    "Time/TPS accuracy", "Time format", "Timer update", "Rolling average length 1 (3 ~ 1000)",
    "Rolling average length 1 (3 ~ 1000)", "Rolling average type 1", "Rolling average type 2", "About", "Hide function button", 
    "Add scramble in details", "Hide summary"],
    ["横向阶数 (2 ~ 20)", "纵向阶数 (2 ~ 20)", "模式", "连拧个数 (> 1)", "多盲个数 (> 1)", "文字颜色",
    "文字大小比例 (0 ~ 1)", "拼图大小 (> 0)", "间隙 (>= 0)", "圆角大小比例 (0 ~ 0.5)", "块的边界", "进制数 (2 ~ 36)",
    "配色方案", "彩虹色亮度",
    "动态配色: 已完成", "当前", "下一组", "其它块", "反转键盘控制方向", "滑动", "语言",
    "时间/TPS 准确度", "时间格式", "计时器刷新", "滚动平均长度1 (3 ~ 1000)",
    "滚动平均长度2 (3 ~ 1000)", "滚动平均类型1", "滚动平均类型2", "关于", "隐藏功能键", 
    "在详情中添加打乱", "隐藏小结"]
];
let colorSchemeList = [
    ["Fringes", "Rows, Columns", "Rows", "Columns", "Mono", "Last 2 Rows", "Last 2 Columns",
    "Live Fringes", "Live Rows, Columns", "Live Rows", "Live Columns", "Live Last 2 Rows", "Live Last 2 Columns"],
    ["降阶", "行列交替", "行", "列", "单色", "按行, 最后2行按列", "按列, 最后2列按行",
    "动态降阶", "动态行列交替", "动态行", "动态列", "动态按行, 最后2行按列", "动态按列, 最后2列按行"]
];
let solveTypeList = [
    ["Standard", "N-2 Relay", "Width Relay", "Height Relay", "Width & Height Relay", "Marathon", "Blindfolded", "Multi Blindfolded"],
    ["标准", "N-2阶递减连拧", "横向阶数递减", "纵向阶数递减", "横向纵向阶数递减", "连拧", "盲拧", "多盲"]
];
let infoBarList = [
    ["Progress ", "Solved ", " Time ", " Moves ", " TPS ", 
    "Current state has been copied to clipboard", "Export is not allowed in the process of blindfolding", 
    "Session mean(", "Enter a session name:"],
    ["进度 ", "复原 ", " 时间 ", " 步数 ", " TPS ", 
    "当前状态已复制到剪切板", "盲拧过程不允许导出状态", 
    "分组平均(", "输入分组名称:"]
];
let resultInteractionList = [
    ["Delete this solve?"], 
    ["确认删除该次成绩?"]
];
let averageBoxList = [
    ["Average", "Mean", "Time", "Moves"], 
    ["去尾平均", "平均", "时间", "步数"]
];
let customizeHintList = [
    ["Enter a scramble:\nInput numbers should be in base 10.\nUse 0 to represent the space.\ne.g. 1 2 3 4/5 6 7 8/9 10 11 12/0 13 14 15",
    "Input puzzle contains unexpected symbols!\nMake sure input contains only numbers, spaces and slashes.",
    "Input puzzle contains unequal pieces per row/column!", "Input puzzle should contain at least 2 rows and 2 columns!",
    "Input puzzle is too big!", "Input puzzle is solved!", "Input puzzle contains duplicate numbers!",
    "Some numbers are too big!", "Input puzzle is not solvable. Continue with this input?",
    "Clear this session?", "Enter a name for new session", "Delete session?\nYou cannot undo this!", 
    "You can't delete your only session."],

    ["请输入一个十进制打乱, 用0代表空格. 例如1 2 3 4/5 6 7 8/9 10 11 12/0 13 14 15",
    "打乱含有其他字符, 请确保打乱只含有数字, 空格和斜杠.",
    "每行/每列的块数不相等!", "打乱应当至少有2行2列!",
    "打乱太大!", "请不要输入复原态!", "打乱含有重复数字!",
    "部分数字太大了!", "该打乱不可解, 要继续吗?",
    "确定清除该分组?", "输入新分组名称:", "删除该分组?\n该操作不可撤销!", 
    "不能删除仅有的分组."]
];
let timerUpdateList = [
    ["real time", "to 0.1s", "to seconds", "do not display"],
    ["实时刷新", "到0.1秒", "到秒", "不刷新"]
];

function setLanguage(systemDefault, languageValue) {
    if (systemDefault) {
        switch (navigator.language.toLowerCase().substr(0, 2)) {
            case "zh":
                language = 1;
                break;
            case "en": default:
                language = 0;
                break;
        }
    } else {
        language = languageValue;
    }
    localStorage.setItem("language", language);

    document.querySelector('#customizeBtn').value = buttonLanguageList[language][0];
    document.querySelector('#scrambleBtn').value = buttonLanguageList[language][1];
    document.querySelector('#solveBtn').value = buttonLanguageList[language][2];
    document.querySelector('#confirmBtn').value = buttonLanguageList[language][3];
    document.querySelector('#nextBtn').value = buttonLanguageList[language][4];
    document.querySelector('#settingsBtn').value = buttonLanguageList[language][5];
    document.querySelector('#backBtn').value = buttonLanguageList[language][6];
    document.querySelector('#exportBtn').value = buttonLanguageList[language][7];
    document.querySelector('#prevBtn').value = buttonLanguageList[language][8];
    document.querySelector('#resultBtn').value = buttonLanguageList[language][9];
    document.querySelector('#clrSession').value = buttonLanguageList[language][10];
    document.querySelector('#renameSession').value = buttonLanguageList[language][11];
    document.querySelector('#newSession').value = buttonLanguageList[language][12];
    document.querySelector('#deleteSession').value = buttonLanguageList[language][13];

    document.querySelector("#statusBar").textContent = infoBarList[language][2] + "0" + infoBarList[language][3] + "0" + infoBarList[language][4] + "0";

    let averageString = ["", ""];
    for (let i = 0; i < 2; i++) {
        if (averageType[i] == 1) {
            averageString[i] = averageString[i] + "Ao" + averageNumber[i];
        } else {
            averageString[i] = averageString[i] + "Mo" + averageNumber[i];
        }
    }
    
    document.querySelector('#titleLabel1').innerHTML = "-----" + settingsTitleList[language][0] + "-----";
    document.querySelector('#titleLabel2').innerHTML = "-----" + settingsTitleList[language][1] + "-----";
    document.querySelector('#titleLabel3').innerHTML = "-----" + settingsTitleList[language][2] + "-----";
    document.querySelector('#titleLabel4').innerHTML = "-----" + settingsTitleList[language][3] + "-----";
    document.querySelector('#titleLabel5').innerHTML = "-----" + settingsTitleList[language][4] + "-----";
    document.querySelector('#titleLabel6').innerHTML = "-----" + settingsTitleList[language][5] + "-----";
    document.querySelector('#titleLabel7').innerHTML = "-----" + settingsTitleList[language][6] + "-----";
    document.querySelector('#resultSessionLabel').innerHTML = settingsTitleList[language][7];

    document.querySelector('#std').innerHTML = solveTypeList[language][0];
    document.querySelector('#tt').innerHTML = solveTypeList[language][1];
    document.querySelector('#tf').innerHTML = solveTypeList[language][2];
    document.querySelector('#ft').innerHTML = solveTypeList[language][3];
    document.querySelector('#ff').innerHTML = solveTypeList[language][4];
    document.querySelector('#mara').innerHTML = solveTypeList[language][5];
    document.querySelector('#bld').innerHTML = solveTypeList[language][6];
    document.querySelector('#multibld').innerHTML = solveTypeList[language][7];

    document.querySelector('#scheme1').innerHTML = colorSchemeList[language][0];
    document.querySelector('#scheme2').innerHTML = colorSchemeList[language][1];
    document.querySelector('#scheme3').innerHTML = colorSchemeList[language][2];
    document.querySelector('#scheme4').innerHTML = colorSchemeList[language][3];
    document.querySelector('#scheme5').innerHTML = colorSchemeList[language][4];
    document.querySelector('#scheme6').innerHTML = colorSchemeList[language][5];
    document.querySelector('#scheme7').innerHTML = colorSchemeList[language][6];
    document.querySelector('#scheme8').innerHTML = colorSchemeList[language][7];
    document.querySelector('#scheme9').innerHTML = colorSchemeList[language][8];
    document.querySelector('#scheme10').innerHTML = colorSchemeList[language][9];
    document.querySelector('#scheme11').innerHTML = colorSchemeList[language][10];
    document.querySelector('#scheme12').innerHTML = colorSchemeList[language][11];
    document.querySelector('#scheme13').innerHTML = colorSchemeList[language][12];

    document.querySelector('#timer1').innerHTML = timerUpdateList[language][0];
    document.querySelector('#timer2').innerHTML = timerUpdateList[language][1];
    document.querySelector('#timer3').innerHTML = timerUpdateList[language][2];
    document.querySelector('#timer4').innerHTML = timerUpdateList[language][3];

    document.querySelector('#averageBox1').innerHTML = averageBoxList[language][0];
    document.querySelector('#meanBox1').innerHTML = averageBoxList[language][1];
    document.querySelector('#averageBox2').innerHTML = averageBoxList[language][0];
    document.querySelector('#meanBox2').innerHTML = averageBoxList[language][1];

    document.querySelector('#puzzleWidthLabel').innerHTML = settingsLanguageList[language][0];
    document.querySelector('#puzzleHeightLabel').innerHTML = settingsLanguageList[language][1];
    document.querySelector('#solveTypeLabel').innerHTML = settingsLanguageList[language][2];
    document.querySelector('#marathonSolvesLabel').innerHTML = settingsLanguageList[language][3];
    document.querySelector('#multibldSolvesLabel').innerHTML = settingsLanguageList[language][4];
    document.querySelector('#textColorLabel').innerHTML = settingsLanguageList[language][5];
    document.querySelector('#textSizeLabel').innerHTML = settingsLanguageList[language][6];
    document.querySelector('#puzzleSizeLabel').innerHTML = settingsLanguageList[language][7];
    document.querySelector('#marginLabel').innerHTML = settingsLanguageList[language][8];
    document.querySelector('#filletLabel').innerHTML = settingsLanguageList[language][9];
    document.querySelector('#borderLabel').innerHTML = settingsLanguageList[language][10];
    document.querySelector('#baseLabel').innerHTML = settingsLanguageList[language][11];
    document.querySelector('#colorSchemeLabel').innerHTML = settingsLanguageList[language][12];
    document.querySelector('#rainbowLightnessLabel').innerHTML = settingsLanguageList[language][13];
    document.querySelector('#liveDoneLabel').innerHTML = settingsLanguageList[language][14];
    document.querySelector('#liveCurrentLabel').innerHTML = settingsLanguageList[language][15];
    document.querySelector('#liveNextLabel').innerHTML = settingsLanguageList[language][16];
    document.querySelector('#liveOtherLabel').innerHTML = settingsLanguageList[language][17];
    document.querySelector('#invertKeyboardLabel').innerHTML = settingsLanguageList[language][18];
    document.querySelector('#switchHoverLabel').innerHTML = settingsLanguageList[language][19];
    document.querySelector('#languageLabel').innerHTML = settingsLanguageList[language][20];
    document.querySelector('#accuracyLabel').innerHTML = settingsLanguageList[language][21];
    document.querySelector('#minutesLabel').innerHTML = settingsLanguageList[language][22];
    document.querySelector('#updateFreqLabel').innerHTML = settingsLanguageList[language][23];
    document.querySelector('#averageNumberLabel1').innerHTML = settingsLanguageList[language][24];
    document.querySelector('#averageNumberLabel2').innerHTML = settingsLanguageList[language][25];
    document.querySelector('#averageTypeLabel1').innerHTML = settingsLanguageList[language][26];
    document.querySelector('#averageTypeLabel2').innerHTML = settingsLanguageList[language][27];
    document.querySelector('#about').innerHTML = settingsLanguageList[language][28];
    document.querySelector('#disableFunctionButtonLabel').innerHTML = settingsLanguageList[language][29];
    document.querySelector('#addScrambleLabel').innerHTML = settingsLanguageList[language][30];
    document.querySelector('#disableSummaryLabel').innerHTML = settingsLanguageList[language][31];
    setDefaultSelected("language", language);

    drawSpacebarCanvas();
    reloadResult();
}